packages/core/src/tracing/instrumentation/protocols/grpcJs.js

```diff
@@ -0,0 +1,504 @@
+      }
+
+      parentSpan.cancel();
+    }
+
+    return cls.ns.runAndReturn(() => {
+      const metadata = call.metadata;
+      const level = readMetadata(metadata, constants.traceLevelHeaderName);
+
+      if (level === '0') {
+        cls.setTracingLevel('0');
+      }
+
+      if (!isActive || cls.tracingSuppressed()) {
+        return originalHandler.apply(originalThis, originalArgs);
+      }
+
+      cls.ns.bindEmitter(call);
+
+      const incomingTraceId = readMetadata(metadata, constants.traceIdHeaderName);
+      const incomingSpanId = readMetadata(metadata, constants.spanIdHeaderName);
+
+      const span = cls.startSpan('rpc-server', constants.ENTRY, incomingTraceId, incomingSpanId);
+      span.data.rpc = {
+        call: dropLeadingSlash(name),
+        flavor: 'grpc'
+      };
+
+      if (TYPES_WITH_CALLBACK.indexOf(type) >= 0) {
+        const originalCallback = originalArgs[1];
+
+        originalArgs[1] = cls.ns.bind(function (err) {
+          if (err) {
+            span.ec = 1;
+            if (err.message || err.details) {
+              span.data.rpc.error = err.message || err.details;
+            }
+          }
+          span.d = Date.now() - span.ts;
+          span.transmit();
+          return originalCallback.apply(this, arguments);
+        });
+      }
+
+      if (TYPES_WITH_CALL_END.indexOf(type) >= 0) {
+        const originalEnd = call.end;
+
+        call.end = function () {
+          span.d = Date.now() - span.ts;
+          process.nextTick(() => {
+            // If the server emits an error, grpc calls call.end before the 'error' event handlers are processed, so we
+            // give on('error') a chance to fire and mark the span erroneous before transmitting it.
+            span.transmit();
+          });
+          return originalEnd.apply(this, arguments);
+        };
+
+        call.on('error', err => {
+          span.ec = 1;
+          if (err.message || err.details) {
+            span.data.rpc.error = err.message || err.details;
+          }
+        });
+
+        call.on('cancelled', () => {
+          span.d = Date.now() - span.ts;
+          span.transmit();
+        });
+      }
+
+      return originalHandler.apply(originalThis, originalArgs);
+    });
+  };
+}
+
+function instrumentedClientMethod(
+  ctx,
+  originalFunction,
+  originalArgs,
+  address,
+  rpcPath,
+  requestStream,
+  responseStream,
+  modifyArgsFn
+) {
+  return cls.ns.runAndReturn(() => {
+    const span = cls.startSpan('rpc-client', constants.EXIT);
+    span.ts = Date.now();
+    span.stack = tracingUtil.getStackTrace(instrumentedClientMethod);
+
+    span.data.rpc = {
+      host: address.host,
+      port: address.port,
+      call: dropLeadingSlash(rpcPath),
+      flavor: 'grpc'
+    };
+
+    modifyArgsFn(originalArgs, span);
+
+    const call = originalFunction.apply(ctx, originalArgs);
+
+    if (requestStream || responseStream) {
+      cls.ns.bindEmitter(call);
+    }
+
+    if (responseStream) {
```

Comment:
_question:_ How do we end the span if responseStream is false? Can we guarantee that there is always a callback wrapped by us for these cases?

Reply:
If false we add the wrapped callback in modifyArgs
You can see [here](https://github.com/instana/nodejs/pull/505/files#diff-d296a9f99c1bb49c23052238bc92ac8a81064b82e55ed06c8f4c403721a15cabR59) where responseStream is false

Reply:
TODO: Check if e.g. makeUnaryCall passes no callback in unit test